providers/common/ciphers/aes_gcm_s390x.inc

```diff
@@ -0,0 +1,327 @@
+        kma->t.g[0] = 0;
+        kma->t.g[1] = 0;
+        /* Set the 32 bit counter */
+        kma->cv.w = kma->j0.w[3];
+    }
+    ctx->ivlen = ivlen;
+}
+
+static int s390x_aes_gcm_cipher_final(PROV_AES_GCM_KEY *ctx, unsigned char *tag)
+{
+    S390X_KMA_PARAMS *kma = &ctx->plat.s390x.param.kma;
+    unsigned char out[16];
+
+    kma->taadl <<= 3;
+    kma->tpcl <<= 3;
+    s390x_kma(ctx->plat.s390x.ares, ctx->plat.s390x.areslen,
+              ctx->plat.s390x.mres, ctx->plat.s390x.mreslen, out,
+              ctx->plat.s390x.fc | S390X_KMA_LAAD | S390X_KMA_LPC, kma);
+
+    /* gctx->mres already returned to the caller */
+    OPENSSL_cleanse(out, ctx->plat.s390x.mreslen);
+    ctx->iv_state = IV_STATE_FINISHED;
+
+    if (ctx->enc) {
+        ctx->taglen = 16;
+        memcpy(tag, kma->t.b, ctx->taglen);
+    } else {
+        if (ctx->taglen < 0)
+            return 0;
+
+        if (CRYPTO_memcmp(tag, kma->t.b, ctx->taglen) != 0)
+            return 0;
+    }
+    return 1;
+}
+
+/*
+ * in_len = length of PlainText & CipherText
+ * For encryption:
+ *   in = Plaintext
+ *   out = CipherText + tag
+ * For decryption:
+ *   in = CipherText + Tag
+ *   out = returned plaintext
+ * tag is the calculated Authentication Tag (with a size of 16 bytes)
+ */
+static int s390x_aes_gcm_one_shot(PROV_AES_GCM_KEY *ctx,
+                                  unsigned char *aad, size_t aad_len,
+                                  const unsigned char *in, size_t in_len,
+                                  unsigned char *out,
+                                  unsigned char *tag, size_t taglen)
+{
+    S390X_KMA_PARAMS *kma = &ctx->plat.s390x.param.kma;
+
+    kma->taadl = aad_len << 3;
+    kma->tpcl = in_len << 3;
+    s390x_kma(aad, ctx->tls_aad_len, in, in_len, out,
+              ctx->plat.s390x.fc | S390X_KMA_LAAD | S390X_KMA_LPC, kma);
+    memcpy(tag, kma->t.b, taglen);
+    return 1;
+}
+
+/*
+ * Process additional authenticated data. Returns 1 on success. Code is
+ * big-endian.
+ */
+static int s390x_aes_gcm_aad_update(PROV_AES_GCM_KEY *ctx,
+                                    const unsigned char *aad, size_t len)
+{
+    S390X_KMA_PARAMS *kma = &ctx->plat.s390x.param.kma;
+    unsigned long long alen;
+    int n, rem;
+
+    /* If already processed pt/ct then error */
+    if (kma->tpcl != 0)
+        return 0;
+
+    /* update the total aad length */
+    alen = kma->taadl + len;
+    if (alen > (U64(1) << 61) || (sizeof(len) == 8 && alen < len))
+        return 0;
+    kma->taadl = alen;
+
+    /* check if there is any existing aad data from a previous add */
+    n = ctx->plat.s390x.areslen;
+    if (n) {
+        /* add additional data to a buffer until it has 16 bytes */
+        while (n && len) {
+            ctx->plat.s390x.ares[n] = *aad;
+            ++aad;
+            --len;
+            n = (n + 1) & 0xf;
+            /* ctx->ares contains a complete block if offset has wrapped around */
+            if (n == 0) {
```

Comment:
If statement was below the loop in original submission. I wonder why one would put more code inside a loop than is needed ?

Reply:
I have changed it back - I meant to take the n out of the while(n && len) {}
The if probably causes an extra branch.